### PR TITLE
feat: add Vault backend

### PR DESCRIPTION
**Pull request by billguo99**
https://app.clickup.com/t/2r1kcha
Moving Nautilus Wallet `web-server` backend to `nautilus-trusted-compute` repository.

#### NW Web Server directory structure
* web-server
  * http-service-impl
  * sgx-helpers
  * sgx-wallet
  * sgx-wallet-impl
  * sgx-wallet-test

#### NTC directory structure
* rust-sgx-workspace
  * projects
    * ntc-tee-server
    * sgx-wallet
    * sgx-wallet-test
  * crates
    * http-service-impl
    * sgx-helpers
    * sgx-wallet-impl

Note: I've excluded `sgx-wallet-impl`, `sgx-wallet`, `sgx-wallet-test` from the rust-sgx-workspace as members because they require certain `[patch]` and I encountered errors  with `std`:
```
error: duplicate lang item in crate `std` (which `base64` depends on): `f32_runtime`.
  |
  = note: the lang item is first defined in crate `sgx_tstd` (which `sgx_wallet_impl` depends on)
  = note: first definition in `sgx_tstd` loaded from /home/bill/workspace/nautilus-trusted-compute/rust-sgx-workspace/target/debug/deps/libsgx_tstd-283919adadf28e07.rmeta
  = note: second definition in `std` loaded from /home/bill/.rustup/toolchains/nightly-2021-11-01-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-8adcca4f1427867b.rlib

error: duplicate lang item in crate `std` (which `base64` depends on): `f64_runtime`.
  |
  = note: the lang item is first defined in crate `sgx_tstd` (which `sgx_wallet_impl` depends on)
  = note: first definition in `sgx_tstd` loaded from /home/bill/workspace/nautilus-trusted-compute/rust-sgx-workspace/target/debug/deps/libsgx_tstd-283919adadf28e07.rmeta
  = note: second definition in `std` loaded from /home/bill/.rustup/toolchains/nightly-2021-11-01-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-8adcca4f1427867b.rlib

error: duplicate lang item in crate `std` (which `base64` depends on): `panic_impl`.
  |
  = note: the lang item is first defined in crate `sgx_tstd` (which `sgx_wallet_impl` depends on)
  = note: first definition in `sgx_tstd` loaded from /home/bill/workspace/nautilus-trusted-compute/rust-sgx-workspace/target/debug/deps/libsgx_tstd-283919adadf28e07.rmeta
  = note: second definition in `std` loaded from /home/bill/.rustup/toolchains/nightly-2021-11-01-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-8adcca4f1427867b.rlib

error: duplicate lang item in crate `std` (which `base64` depends on): `begin_panic`.
  |
  = note: the lang item is first defined in crate `sgx_tstd` (which `sgx_wallet_impl` depends on)
  = note: first definition in `sgx_tstd` loaded from /home/bill/workspace/nautilus-trusted-compute/rust-sgx-workspace/target/debug/deps/libsgx_tstd-283919adadf28e07.rmeta
  = note: second definition in `std` loaded from /home/bill/.rustup/toolchains/nightly-2021-11-01-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-8adcca4f1427867b.rlib

error: duplicate lang item in crate `std` (which `base64` depends on): `oom`.
  |
  = note: the lang item is first defined in crate `sgx_tstd` (which `sgx_wallet_impl` depends on)
  = note: first definition in `sgx_tstd` loaded from /home/bill/workspace/nautilus-trusted-compute/rust-sgx-workspace/target/debug/deps/libsgx_tstd-283919adadf28e07.rmeta
  = note: second definition in `std` loaded from /home/bill/.rustup/toolchains/nightly-2021-11-01-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-8adcca4f1427867b.rlib
```

JP mentioned a possible solution is to use `xargo` to replace `cargo`.

### CODE REVIEW

**Review by IscoRuta98 — APPROVED** [review]

**Comment by binglekruger** [comment]
Update enclave memory to ~2 GB for now

```diff
diff --git a/rust-sgx-workspace/projects/ntc-tee-server/enclave/Enclave.config.xml b/rust-sgx-workspace/projects/ntc-tee-server/enclave/Enclave.config.xml
index ee4c3f7..bf34e90 100644
--- a/rust-sgx-workspace/projects/ntc-tee-server/enclave/Enclave.config.xml
+++ b/rust-sgx-workspace/projects/ntc-tee-server/enclave/Enclave.config.xml
@@ -2,8 +2,8 @@
 <EnclaveConfiguration>
   <ProdID>0</ProdID>
   <ISVSVN>0</ISVSVN>
-  <StackMaxSize>0x40000</StackMaxSize>
-  <HeapMaxSize>0x100000</HeapMaxSize>
+  <StackMaxSize>0x8000000</StackMaxSize>
+  <HeapMaxSize>0x80000000</HeapMaxSize>
   <TCSNum>1</TCSNum>
   <TCSPolicy>1</TCSPolicy>
   <DisableDebug>0</DisableDebug>
```

**Review by binglekruger — COMMENTED** [review]
I skipped `projects/sgx-wallet-test` and `projects/sgx-wallet`

**Comment by billguo99** [comment]
> Update enclave memory to ~2 GB for now

That `Enclave.config.xml` is for `ntc-tee-server` which I believe will be the Trusted Execution Environment for the WASM runtime in the future. Once we've done with experimentation on `wasm-exec-sgx`, we can change these files based on the `wasm-exec-sgx` repo.

File: rust-sgx-workspace/crates/sgx-wallet-impl/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+# Community SGX forks
+rand = { git = "https://github.com/mesalock-linux/rand-sgx" }
+rmp-serde = { git = "https://github.com/mesalock-linux/msgpack-rust-sgx" }
+serde = { git = "https://github.com/mesalock-linux/serde-sgx" }
+serde_json = { git = "https://github.com/mesalock-linux/serde-json-sgx" }
+thiserror = { git = "https://github.com/mesalock-linux/thiserror-sgx" }
+
+# Our SGX forks
+algonaut = { git = "https://github.com/registreerocks/algonaut-sgx", branch = "main-sgx" }
+serde_bytes = { version = "0.11.4", git = "https://github.com/registreerocks/serde-bytes-sgx" } # SGX: registreerocks fork for 0.11.4
```
[the remark this answers is not in the document]
Comment (billguo99) [reply rust-sgx-workspace/crates/sgx-wallet-impl/Cargo.toml:35]: Good point, I'll create a backlog ticket for this as if we fork these registree's repos, we should change them for Nautilus Wallet as well.

[the remark this answers is not in the document]
Comment (billguo99) [reply rust-sgx-workspace/crates/sgx-wallet-impl/Cargo.toml:35]: https://app.clickup.com/t/2tn2g6p

**Pull request merged**